src/main/java/org/veupathdb/service/eda/ds/plugin/CategoricalCountsPlugin.java

```diff
@@ -0,0 +1,140 @@
+import org.veupathdb.service.eda.ds.Resources;
+import org.veupathdb.service.eda.ds.plugin.AbstractEmptyComputePlugin;
+import org.veupathdb.service.eda.generated.model.APIVariableType;
+import org.veupathdb.service.eda.generated.model.APIVariableDataShape;
+import org.veupathdb.service.eda.generated.model.CategoricalCountsSpec;
+import org.veupathdb.service.eda.generated.model.CategoricalCountsPostRequest;
+import org.veupathdb.service.eda.generated.model.CategoricalCountsPostResponse;
+import org.veupathdb.service.eda.generated.model.VariableSpec;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.gusdb.fgputil.FormatUtil.TAB;
+
+public class CategoricalCountsPlugin extends AbstractEmptyComputePlugin<CategoricalCountsPostRequest, CategoricalCountsSpec> {
+  
+  private static final Logger LOG = LogManager.getLogger(CategoricalCountsPlugin.class);
+
+  private static final ManagedMap<String,CategoricalCountsPostResponse> RESULT_CACHE =
+      new ManagedMap<>(5000, 2000);
+
+  private String _cacheKey;
+  private CategoricalCountsPostResponse _cachedResponse;
+
+  @Override
+  protected Class<CategoricalCountsPostRequest> getVisualizationRequestClass() {
+    return CategoricalCountsPostRequest.class;
+  }
+
+  @Override
+  protected Class<CategoricalCountsSpec> getVisualizationSpecClass() {
+    return CategoricalCountsSpec.class;
+  }
+
+  @Override
+  public ConstraintSpec getConstraintSpec() {
+    return new ConstraintSpec()
+      .dependencyOrder(List.of("variable"))
+      .pattern()
+        .element("variable")
+          .shapes(APIVariableDataShape.CATEGORICAL, APIVariableDataShape.ORDINAL)
+          .description("Variable must be categorical.")
+      .done();
+  }
+  
+  @Override
+  protected void validateVisualizationSpec(CategoricalCountsSpec pluginSpec) throws ValidationException {
+    validateInputs(new DataElementSet()
+      .var("variable", pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected List<StreamSpec> getRequestedStreams(CategoricalCountsSpec pluginSpec) {
+    String entityId = pluginSpec.getVariable().getEntityId();
+    _cacheKey = EncryptionUtil.md5(
+        JsonUtil.serializeObject(getSubsetFilters()) +
+        "|" + JsonUtil.serializeObject(pluginSpec)
+    );
+    _cachedResponse = RESULT_CACHE.get(_cacheKey);
+    if (_cachedResponse != null) LOG.info("Found cached result.");
+
+    return _cachedResponse != null ? Collections.emptyList() : ListBuilder.asList(
+      new StreamSpec(DEFAULT_SINGLE_STREAM_NAME, pluginSpec.getVariable().getEntityId())
+        .addVar(pluginSpec.getVariable()));
+  }
+
+  @Override
+  protected void writeResults(OutputStream out, Map<String, InputStream> dataStreams) throws IOException {
+    CategoricalCountsSpec spec = getPluginSpec();
```

Comment:
You're never adding the result to the cache here.  The condition on line 83 says: "if we found the result in the cache, we don't need to request a stream", but then in writeResults() you read the stream anyway, so that would give you a NullPointerException on 93.  The trick is to check again if _cachedReponse is null (that's why it and _cacheKey are fields, not local vars), and return if so.  Otherwise you have to call RESULT_CACHE.getValue() where you essentially tell the cache how to create the value to be put into the cache.  This API is so the creation+insert+return happens atomically and is threadsafe.  See [the impl here](https://github.com/VEuPathDB/EdaDataService/blob/master/src/main/java/org/veupathdb/service/eda/ds/plugin/sample/RecordCountPlugin.java) for an example.